include/LightGBM/utils/yamc/alternate_shared_mutex.hpp

```diff
@@ -0,0 +1,212 @@
+/*
+ * alternate_shared_mutex.hpp
+ *
+ * MIT License
+ *
+ * Copyright (c) 2017 yohhoy
```

Comment:
What is this copyright?  Did this file come from a different repository?
" * Copyright (c) 2017 yohhoy"

Reply:
Yes, it comes from https://github.com/yohhoy/yamc

Reply:
looks like it comes from here:
https://github.com/yohhoy/yamc
Hmm, is there a way to get the built artifacts from that library instead of copying some of the files over?  Ideally we wouldn't want to create duplicate code (or a "fork" in some sense) - it would be better to update the mutex library code in one place (eg for bug fixes, new features, etc).

Oh, what about git submodules?  Maybe that could be used here?
Example link:
https://git-scm.com/book/en/v2/Git-Tools-Submodules

Reply:
Since it is a header-only library, I think they do not have any built artifacts.

Reply:
the ideal soloution I was proposing in this PR was having the possibility to use c++17, but too much trouble would be caused. And coming with a custom implementation of shared mutex and locks is really error-prone.